### ISSUE TITLE
Is there are any plans to optimise for neon arm64 instructions? 

### ISSUE THREAD

**Issue by freedbrt** [tvlabs/edge264#9]
This project is optimised only for SSE3 instructions? Are there any benchmarks for arm64 processors?

**Comment by traffaillac** [comment]
Yes! At the moment I am busy with future job applications as academic researcher, then I'll be unemployed until September and will work on making edge264 as production-ready as can be :
* multithreading first since it impacts API and I want it frozen sooner
* integration in ffmpeg/VLC
* ARM support (both 32 and 64 bit ideally)
The project is optimized for SSSE3, with a few key speedups for SSE4.2 and AVX2. Unlike other decoders, I do not maintain many versions of each SIMD routine (e.g. for SSE2, SSSE3, AVX2, AVX512) to reduce maintenance nightmares (which do not bring much speedup anyway). So ARM support will certainly be limited to NEON instead of SVE. Unless I can find a programming language that allows writing one code version and compile it to perfect NEON and SVE assembly :)

**Comment by traffaillac** [comment]
ARM64 is now complete, I'm marking this issue as resolved :)

As for ISPC, I do not intend to use it so far. I have looked recently at ways to complement edge264 with a portable fallback for less mainstream architectures, and currently intend to go for Clang vector extensions (__builtin_shufflevector, __builtin_convertvector, __builtin_elementwise, __builtin_reduce) as I really like their minimalism and explicit use of vector operations. All my code is already expressed with vector operations so it would make little sense to express it back in scalar operations just to let ISPC vectorize it back. In general I am not a proponent of auto-vectorization solutions, I really want something with explicit operations but could not see them in the docs. It would really help to have a simpler doc like https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html !
I guess a good candidate for ISPC could be GPU offloading of the deblocking filter, the same way Dav1d demonstrated with AV1, but that's not on my plans yet. Forks are welcome to experiment on that :)

Cheers,
Thibault

**Comment by rnvannatta** [comment]
Hey Thibault,

I use ISPC for Civilization 7 to ship sse, avx, avx2 and arm neon. ISPC isn't autovectorization, it was specifically designed because autovectorization is not very good. It's more like a shader language, where you have a workgroup (ispc calls this a gang) and individual threads (ispc calls this an instance).

Take the code:

```C
foreach(i = 0 ... ct) {
  sum[i] = a[i] + b[i]
}
```

this is _guaranteed_ to compile to vector instructions because ISPC programs are vector programs, and when compiled to SSE is roughly equivalent to

```C
int i = 0;
for(; i < ct / 4; i+=4) {
  _mm_store_ps(sum + i*4, _mm_add_ps(_mm_loadu_ps(a+i*4), _mm_loadu_ps(b+i*4));
}
for(i = i-4; i < ct; i++) {
  sum[i] = a[i] + b[i];
}
```

[you can also explicitly loop in increments of programCount](https://ispc.github.io/ispc.html#parallel-iteration-with-programindex-and-programcount)
